docs/docs/canonicai/chapter3.md

```diff
@@ -0,0 +1,339 @@
+```
+node cai_state {
+    has name;
+    has prepared_entities = {
+            "name": "Tony Stark",
+            "address": "10880 Malibu Point"
+    };
+
+    # Classify the intent of the question
+    can classify_intent {
+        # NOTE: Hardcode intent for now
+        visitor.predicted_intent = visitor.intent_override;
+    }
+
+    # Extract words of interests from the question
+    can extract_entities {
+        # NOTE: Hardcode entities for now
+        for ent in visitor.entity_override {
+            visitor.extracted_entities[ent] = prepared_entities[ent];
+        }
+    }
+
+    # Generate response
+    can gen_response {
+        # Default response
+        visitor.response =
+            "Sorry I can't handle that just yet! Anything else I can help you with";
+    }
+
+    # Process incoming request through NLU engines
+    can nlu {
+        ::classify_intent;
+        ::extract_entities;
+    }
+
+    # Construct natural language response
+    can nlg {
+        ::gen_response;
+    }
+}
+```
+Above, this is a conversational state node. We use inheritance here, because we will have multiple conversational state nodes like a confirm state, cancel state, collect information state and etc, that have the same functionality but does have different node abilities.
+
+```
+node collect_info:cai_state {
+    has name = "collect_info";
+    can gen_response {
+        if ("name" in visitor.extracted_entities and
+                "address" not in visitor.extracted_entities):
+            visitor.response = "What is your address?"
+        elif ("address" in visitor.extracted_entities and
+                "name" not in visitor.extracted_entities):
+            visitor.response = "What is your name?"
+        else:
+            visitor.response =
+                "To set you up with a test drive, we will need your name and address.";
+    }
+}
+
+node confirmation:cai_state {
+    has name = "confirmation";
+    can gen_response {
+        visitor.response =
+            "Can you confirm your name to be " + visitor.extracted_entities["name"] + " and your address as " + visitor.extracted_entities["address"] + " ?";
+    }
+}
+
+node confirmed:cai_state {
+    has name = "confirmed";
+    can gen_response {
+        visitor.response = "You are all set for a Tesla test drive!";
+    }
+}
+
+node canceled:cai_state {
+    has name = "canceled";
+    can gen_response {
+        visitor.response = "No worries. We look forward to hear from you in the future!";
+    }
+}
+```
+As you can see here, all of the conversational state node have something in common, that's why it is very important to use inheritance, because this code instead of being 200 lines of code it can be over 1000+ lines of code and it could be very hard to manage and read.
+
+You can go to [jaseci docs](https://docs.jaseci.org/) for more information on inheritance. If you are still confused about inheritance and how it works.
```

Comment:
Can we link to the specific page/section that talks about inheritance? If we don't have that in the docs yet, we can mention specific chapter of the bible.

This applies to all the places we want to link people to documentation throughout these codelabs.

Reply:
It's like a single page application. It don't have a link to the exact inheritance section.

Reply:
Not sure if that's true. I can certainly link to a specific page in the docs, for example, https://docs.jaseci.org/docs/getting-started/JAC-Language-Overview/report.

Now, I am not sure if the current docs have any page on inheritance. If it doesn't, that's a different story.

Reply:
Yes now I see it. It was under this link.

https://docs.jaseci.org/docs/Developing_with_JAC/Language_Features/OOP